java/src/main/native/src/RmmJni.cpp

```diff
@@ -344,44 +347,42 @@ JNIEXPORT void JNICALL Java_ai_rapids_cudf_Rmm_initializeInternal(JNIEnv *env, j
     bool use_pool_alloc = allocation_mode & 1;
     bool use_managed_mem = allocation_mode & 2;
     bool use_arena_alloc = allocation_mode & 4;
+    bool use_aligned_adapter = allocation_mode & 8;
     if (use_pool_alloc) {
       auto pool_limit = (max_pool_size > 0) ?
                             thrust::optional<std::size_t>{static_cast<std::size_t>(max_pool_size)} :
                             thrust::nullopt;
       if (use_managed_mem) {
         Initialized_resource = rmm::mr::make_owning_wrapper<rmm::mr::pool_memory_resource>(
             std::make_shared<rmm::mr::managed_memory_resource>(), pool_size, pool_limit);
-        auto wrapped = make_tracking_adaptor(Initialized_resource.get(), RMM_ALLOC_SIZE_ALIGNMENT);
-        Tracking_memory_resource.reset(wrapped);
       } else {
         Initialized_resource = rmm::mr::make_owning_wrapper<rmm::mr::pool_memory_resource>(
             std::make_shared<rmm::mr::cuda_memory_resource>(), pool_size, pool_limit);
-        auto wrapped = make_tracking_adaptor(Initialized_resource.get(), RMM_ALLOC_SIZE_ALIGNMENT);
-        Tracking_memory_resource.reset(wrapped);
       }
     } else if (use_arena_alloc) {
       std::size_t pool_limit = (max_pool_size > 0) ? static_cast<std::size_t>(max_pool_size) :
                                                      std::numeric_limits<std::size_t>::max();
       if (use_managed_mem) {
         Initialized_resource = rmm::mr::make_owning_wrapper<rmm::mr::arena_memory_resource>(
             std::make_shared<rmm::mr::managed_memory_resource>(), pool_size, pool_limit);
-        auto wrapped = make_tracking_adaptor(Initialized_resource.get(), RMM_ALLOC_SIZE_ALIGNMENT);
-        Tracking_memory_resource.reset(wrapped);
       } else {
         Initialized_resource = rmm::mr::make_owning_wrapper<rmm::mr::arena_memory_resource>(
             std::make_shared<rmm::mr::cuda_memory_resource>(), pool_size, pool_limit);
-        auto wrapped = make_tracking_adaptor(Initialized_resource.get(), RMM_ALLOC_SIZE_ALIGNMENT);
-        Tracking_memory_resource.reset(wrapped);
       }
     } else if (use_managed_mem) {
       Initialized_resource = std::make_shared<rmm::mr::managed_memory_resource>();
-      auto wrapped = make_tracking_adaptor(Initialized_resource.get(), RMM_ALLOC_SIZE_ALIGNMENT);
-      Tracking_memory_resource.reset(wrapped);
     } else {
       Initialized_resource = std::make_shared<rmm::mr::cuda_memory_resource>();
-      auto wrapped = make_tracking_adaptor(Initialized_resource.get(), RMM_ALLOC_SIZE_ALIGNMENT);
-      Tracking_memory_resource.reset(wrapped);
     }
+
+    if (use_aligned_adapter) {
+      Initialized_resource = rmm::mr::make_owning_wrapper<rmm::mr::aligned_resource_adaptor>(
+          Initialized_resource, allocation_alignment, alignment_threshold);
+    }
+
+    auto wrapped = make_tracking_adaptor(Initialized_resource.get(), RMM_ALLOC_SIZE_ALIGNMENT);
```

Comment:
`RMM_ALLOC_SIZE_ALIGNMENT` seems wrong if we know we're using the aligned adapter and a different alignment.  I think there needs to be a `max(RMM_ALLOC_SIZE_ALIGNMENT, allocation_alignment)` or something similar here.

Reply:
Hmm, the actual allocation size for aligned adapter is a bit complicated. Just curious, why are we tracking the total allocation size ourselves and not using the `get_mem_info()` method from the device memory resource?

Reply:
Because many device memory resource implementations do not implement `get_mem_info()`, the ARENA allocator apparently being one of them:
```
$ git grep supports_get_mem_info | grep false
benchmarks/utilities/simulated_memory_resource.hpp:  bool supports_get_mem_info() const noexcept override { return false; }
include/rmm/mr/device/arena_memory_resource.hpp:  bool supports_get_mem_info() const noexcept override { return false; }
include/rmm/mr/device/binning_memory_resource.hpp:  bool supports_get_mem_info() const noexcept override { return false; }
include/rmm/mr/device/cuda_async_memory_resource.hpp:  bool supports_get_mem_info() const noexcept override { return false; }
include/rmm/mr/device/fixed_size_memory_resource.hpp:  bool supports_get_mem_info() const noexcept override { return false; }
include/rmm/mr/device/pool_memory_resource.hpp:  bool supports_get_mem_info() const noexcept override { return false; }
```

Reply:
Yeah I guess that api is really only for cuda. Added the max of two alignment sizes.